.circleci/build_image.sh

```diff
@@ -0,0 +1,48 @@
+#!/bin/sh
+set -xe
+
+if [[ $# -ne 2 ]]; then
+  echo "Usage: $0 DOCKERFILE IMAGE_NAME"
+  echo ".dockerfile is appended to DOCKERFILE."
+  echo "janitortechnology/ is prepended to IMAGE_NAME."
+  exit 2
+fi
+
+DIR=`dirname $1`
+DOCKERFILE=`basename $1`
+IMAGE_NAME=$2
+
+if [[ -z $CIRCLE_PULL_REQUEST ]]; then
+  # Log in to Docker Hub.
+  docker login -u $DOCKER_USER -p $DOCKER_PASS
+  # Decide whether Docker Hub images should be tagged ":branch-Y" or ":latest".
+  if [[ $CIRCLE_BRANCH != "master" ]]; then
+    DOCKERHUB_TAG="branch-$CIRCLE_BRANCH"
+  else
+    DOCKERHUB_TAG="latest"
+  fi
+else
+  DOCKERHUB_TAG="pr"
+fi
+
+# Pull published image for cached layers. (Allowed to fail)
+docker pull janitortechnology/$IMAGE_NAME:latest || true
+if [[ $CIRCLE_JOB != "ubuntu-dev" ]]; then
+  # Load the `ubuntu-dev` base image from previous step.
+  docker load -i ubuntu-dev.tar
+  if [[ $DOCKERHUB_TAG != "latest" ]]; then
+    docker tag janitortechnology/ubuntu-dev:$DOCKERHUB_TAG janitortechnology/ubuntu-dev:latest
+  fi
+fi
+
+cd "$DIR"
+docker build --cache-from=janitortechnology/$IMAGE_NAME:latest -t janitortechnology/$IMAGE_NAME:$DOCKERHUB_TAG -f "$DOCKERFILE.dockerfile" .
+if [[ -z $CIRCLE_PULL_REQUEST ]]; then
+  docker push $IMAGE_NAME:$DOCKERHUB_TAG
+fi
+cd ..
+
+if [[ $CIRCLE_JOB = "ubuntu-dev" ]]; then
+  # Save the `ubuntu-dev` base image for next step.
+  docker save janitortechnology/$IMAGE_NAME:$DOCKERHUB_TAG -o ubuntu-dev.tar
+fi
```

Comment:
Nit: Please remove this `if [[ $CIRCLE_JOB = "ubuntu-dev" ]]` block from the common script, and add the `docker save` instruction directly to the `ubuntu-dev` job in `circle.yml` (no point in having it in a script that's common to all jobs).